File: firmware/src/core/encryption.cpp
#include "encryption.h"

static uint32_t current_key = 0;

void encryption_init(uint32_t session_key) {
    current_key = session_key;
}

uint32_t encryption_derive_key(uint32_t nonce) {
    return SHARED_SECRET ^ nonce;
}

void encryption_process(uint8_t *data, size_t length) {
    if (current_key == 0) return; // No encryption if key is 0

    for (size_t i = 0; i < length; i++) {
        uint8_t key_byte = (current_key >> ((i % 4) * 8)) & 0xFF;
        data[i] ^= key_byte;
        
        // Evolve key periodically or per byte? Requirements say "Key evolves per packet".
        // Let's evolve it after each full packet processing.
    }
    
    // Evolve key for next packet
    current_key = (current_key * 1103515245 + 12345);
}

File: service/encryption.py
SHARED_SECRET = 0xDEADC0DE

class Encryption:
    def __init__(self, session_key=0):
        self.current_key = session_key

    def derive_key(self, nonce):
        return SHARED_SECRET ^ nonce

    def process(self, data):
        if self.current_key == 0:
            return data
            
        data = bytearray(data)
        for i in range(len(data)):
            key_byte = (self.current_key >> ((i % 4) * 8)) & 0xFF
            data[i] ^= key_byte
            
        # Evolve key
        self.current_key = (self.current_key * 1103515245 + 12345) & 0xFFFFFFFF
        return bytes(data)
